src/lib/crypto/kimchi_backend/gadgets/not.ml

```diff
@@ -0,0 +1,119 @@
+open Core_kernel
+
+open Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint
+
+module Bignum_bigint = Snarky_backendless.Backend_extended.Bignum_bigint
+
+(* NOT *)
+
+(* Boolean Not of length bits for checked length (uses Xor gadgets inside) *)
+let bnot_checked (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (input : Circuit.Field.t) (length : int) (len_xor : int) : Circuit.Field.t =
+  let open Circuit in
+  let all_ones = Bignum_bigint.(pow (of_int 2) (of_int length) - one) in
+  let all_ones = Common.bignum_bigint_to_field (module Circuit) all_ones in
+  let all_ones = Field.constant all_ones in
+
+  (* Negating is equivalent to XORing with all one word *)
+  Xor.bxor (module Circuit) input all_ones length len_xor
+
+(* Boolean Not of length bits for unchecked length (uses Generic subtractions inside) *)
+let bnot_unchecked (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (input : Circuit.Field.t) (length : int) : Circuit.Field.t =
+  let open Circuit in
+  let all_ones = Bignum_bigint.(pow (of_int 2) (of_int length) - one) in
+  let all_ones = Common.bignum_bigint_to_field (module Circuit) all_ones in
+  let all_ones = Field.constant all_ones in
+  let all_ones_var =
```

Comment:
If you do this often, could be nice to have a helper in Common?

Reply:
made one inside not